gradio/route_utils.py

```diff
@@ -583,3 +584,54 @@ def starts_with_protocol(string: str) -> bool:
     """
     pattern = r"^[a-zA-Z][a-zA-Z0-9+\-.]*://"
     return re.match(pattern, string) is not None
+
+
+def get_core_url(original_url: str) -> str:
+    """
+    Returns the core url of a given url. If the url does not have a scheme, it is assumed to be http.
+    Examples:
+        get_core_url("https://www.gradio.app") -> "www.gradio.app"
+        get_core_url("localhost:7860") -> "localhost"
+        get_core_url("127.0.0.1") -> "127.0.0.1"
+    """
+    if not original_url:
+        return ""
+    if "://" not in original_url:
+        original_url = "http://" + original_url
+    return httpx.URL(original_url).host
+
+
+class CustomCORSMiddleware(BaseHTTPMiddleware):
+    async def dispatch(self, request: fastapi.Request, call_next):
+        host_header: str = request.headers.get("host", "")
+        origin_header: str = request.headers.get("origin", "")
+        host_header_host = get_core_url(host_header)
+        origin_header_host = get_core_url(origin_header)
+
+        localhost_aliases = ["localhost", "127.0.0.1", "0.0.0.0", "null"]
+
+        if (
+            host_header_host in localhost_aliases
+            and origin_header_host
+            and origin_header_host not in localhost_aliases
+        ):
+            return fastapi.Response(
+                content="Blocked by CORS policy.",
+                status_code=fastapi.status.HTTP_403_FORBIDDEN,
+            )
+
+        if request.method == "OPTIONS":
```

Comment:
Please look at how the base CORSMiddleware does this. Not all `OPTIONS` requests are CORS preflight requests.

Reply:
Will take a look